Source/WebGPU/WGSL/EntryPointRewriter.cpp

```diff
@@ -0,0 +1,284 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "EntryPointRewriter.h"
+
+#include "AST.h"
+#include "ASTVisitor.h"
+
+namespace WGSL {
+
+class EntryPointRewriter {
+public:
+    EntryPointRewriter(AST::ShaderModule& shaderModule, AST::FunctionDecl& functionDecl)
+        : m_shaderModule(shaderModule)
+        , m_functionDecl(functionDecl)
+        , m_emptySourceSpan(0, 0, 0, 0)
+    {
+    }
+
+    void rewrite();
+
+private:
+    struct MemberOrParameter {
+        String m_name;
+        Ref<AST::TypeDecl> m_type;
+        AST::Attribute::List m_attributes;
+    };
+
+    enum class IsBuiltin {
+        No = 0,
+        Yes = 1,
+    };
+
+    static AST::TypeDecl& getResolvedType(AST::TypeDecl&);
+
+    void collectParameters();
+    void constructInputStruct();
+    void materialize(Vector<String>& path, MemberOrParameter&, IsBuiltin);
+    void visit(Vector<String>& path, MemberOrParameter&&);
+    void appendBuiltins();
+
+    AST::ShaderModule& m_shaderModule;
+    AST::FunctionDecl& m_functionDecl;
+
+    const SourceSpan m_emptySourceSpan;
+    Vector<MemberOrParameter> m_builtins;
+    Vector<MemberOrParameter> m_parameters;
+    AST::Statement::List m_materializations;
+    String m_structTypeName;
+    String m_structParameterName;
+};
+
+AST::TypeDecl& EntryPointRewriter::getResolvedType(AST::TypeDecl& type)
+{
+    if (type.kind() == AST::Node::Kind::TypeReference)
+        return getResolvedType(downcast<AST::TypeReference>(type).type());
+
+    if (type.kind() == AST::Node::Kind::NamedType) {
+        if (auto* resolvedType = downcast<AST::NamedType>(type).maybeResolvedReference())
+            return getResolvedType(*resolvedType);
+    }
+
+    return type;
+}
+
+void EntryPointRewriter::rewrite()
+{
+    m_structTypeName = makeString("__", m_functionDecl.name(), "_inT");
+    m_structParameterName = makeString("__", m_functionDecl.name(), "_in");
```

Comment:
We should make sure that m_functionDecl.name() is the mangled name, rather than the name in the original source. Also, how are we going to guarantee there won't be any name collisions?

Reply:
That's the nice thing about doing this as an AST pass. As long as the mangling pass runs after this, we don't have to worry about that. As for the name collisions, the spec says that there can't be two functions with the same name, and the `__` prefix is invalid for user defined values, so I think this should work?